sklearn/preprocessing/tests/test_knn_imputation.py

```diff
@@ -0,0 +1,432 @@
+
+
+def test_knn_imputation_shape():
+    # Verify the shapes of the imputed matrix for different weights and
+    # number of neighbors.
+    n_rows = 10
+    n_cols = 2
+    X = np.random.rand(n_rows, n_cols)
+    X[0, 0] = np.nan
+
+    for weights in ['uniform', 'distance']:
+        for n_neighbors in range(1, 6):
+            imputer = KNNImputer(n_neighbors=n_neighbors, weights=weights)
+            X_imputed = imputer.fit_transform(X)
+            assert_equal(X_imputed.shape, (n_rows, n_cols))
+
+
+def test_knn_imputation_zero():
+    # Test imputation when missing_values == 0
+    missing_values = 0
+    n_neighbors = 2
+    imputer = KNNImputer(missing_values=missing_values,
+                         n_neighbors=n_neighbors,
+                         weights="uniform")
+    # imputer_nan = KNNImputer(missing_values=np.nan,
+    #                          n_neighbors=n_neighbors,
+    #                          weights="uniform")
+
+    # Test with missing_values=0 when NaN present
+    X = np.array([
+        [np.nan, 0, 0, 0, 5],
+        [np.nan, 1, 0, np.nan, 3],
+        [np.nan, 2, 0, 0, 0],
+        [np.nan, 6, 0, 5, 13],
+    ])
+    assert_raises(ValueError, imputer.fit, X)
+
+    # Test with % zeros in column > col_max_missing
+    X = np.array([
+        [1, 0, 0, 0, 5],
+        [2, 1, 0, 2, 3],
+        [3, 2, 0, 0, 0],
+        [4, 6, 0, 5, 13],
+    ])
+    assert_raises(ValueError, imputer.fit, X)
+
+    # Test with an imputable matrix and also compare with missing_values="NaN"
+    # X = np.array([
+    #     [1, 0, 1, 0, 5],
+    #     [2, 1, 2, 2, 3],
+    #     [3, 2, 3, 0, 0],
+    #     [6, 6, 0, 5, 13],
+    # ])
+
+    X = np.array([
+        [1, 0, 1, 0, 1],
+        [2, 1, 2, 2, 3],
+        [3, 2, 3, 0, 0],
+        [6, 6, 0, 5, 17],
+    ])
+
+    # X_nan = np.array([
+    #     [1, np.nan, 1, np.nan, 1],
+    #     [2, 1, 2, 2, 3],
+    #     [3, 2, 3, np.nan, np.nan],
+    #     [6, 6, np.nan, 5, 17],
+    # ])
+
+    statistics_mean = [3, 3, 2, 3.5, 7]
+    X_imputed = np.array([
+        [1, 1.5, 1,   2, 1],
+        [2, 1,   2,   2, 3],
+        [3, 2,   3,   2, 2],
+        [6, 6,   1.5, 5, 17],
+    ])
+
+    assert_array_equal(imputer.fit_transform(X), X_imputed)
+    assert_array_equal(imputer.statistics_, statistics_mean)
+    # The following fails at the moment as NearestNeighbors object does not
+    # pass missing_values=0 to pairwise_distances()
+    # assert_array_equal(imputer.fit_transform(X), imputer_nan.fit_transform(
+    #     X_nan))
+
+
+def test_knn_imputation_default():
+    # Test imputation with default values
+    # imputer = KNNImputer()
+
+    # Test with % missing in a column > col_max_missing
+    X = np.array([
+        [np.nan, 0, 0, 0, 5],
+        [np.nan, 1, 0, np.nan, 3],
+        [np.nan, 2, 0, 0, 0],
+        [np.nan, 6, 0, 5, 13],
+        [np.nan, 7, 0, 7, 8],
+        [np.nan, 8, 0, 8, 9],
+    ])
+    assert_raises(ValueError, KNNImputer().fit, X)
```

Comment:
I think it would be better to test that the right message is raised for something as generic as ValueError. Use `assert_raise_message` or `assert_raises_regexp`